pip/qsharp/_qsharp.py

```diff
@@ -93,23 +93,60 @@ def eval_file(path: str) -> Any:
     return eval(f.read())
 
 
-def run(entry_expr: str, shots: int) -> Any:
+class ShotResult(TypedDict):
+    """
+    A single result of a shot.
+    """
+
+    events: List[Output]
+    result: Any
+
+
+def run(
+    entry_expr: str,
+    shots: int,
+    on_result: Optional[Callable[[ShotResult], None]] = None,
+    save_output: bool = False,
+) -> Any:
```

Comment:
Shall we require the new arguments to be named? I _think_ that gives us more flexibility to change this signature down the line. 

```suggestion
def run(
    entry_expr: str,
    shots: int,
    *,
    on_result: Optional[Callable[[ShotResult], None]] = None,
    save_output: bool = False,
) -> Any:
```